h2/src/main/org/h2/command/Parser.java

```diff
@@ -2483,17 +2486,13 @@ private Query parseQueryExpression() {
         QueryScope outerQueryScope = queryScope;
         Query query;
         if (readIf(WITH)) {
-            queryScope = new QueryScope(outerQueryScope);
+            boolean isPotentiallyRecursive = readIf("RECURSIVE");
```

Comment:
why potentially? I'm ok with making it a requirement for this keyword to be correctly used

Reply:
This is a definition from the SQL Standard. When `RECURSIVE` is specified, a with list and its elements are **potentially recursive**, when it isn't specified they are **non-recursive**.

Elements of a potentially recursive list aren't required to be recursive, but they have the same restrictions as recursive (such as mandatory column names).